Add Context Manager for Disabling Multithreading in Backwards, use in aot autograd

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86245

We were running into a few issues with running multithreaded backwards in aot_autograd: such as https://github.com/pytorch/pytorch/issues/86136, and `FakeTensorMode` getting into a weird state as a result of not executing functions completely sequentially. The multithreaded backwards is lost in translation when we trace out the backwards anyway, and adds a lot of additional complexity.

eellison committed:
Add Context Manager for Disabling Multithreading in Backwards, use in aot autograd

[ghstack-poisoned]

eellison commented on torch/autograd/grad_mode.py:
```diff
@@ -298,3 +298,39 @@ def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+
+    Example::
+        >>> # xdoctest: +SKIP
+        >>> TODO - not sure what python short example would be useful here
```
not sure what i should put here..

ezyang replied:
don't bother

ezyang commented on torch/csrc/autograd/init.cpp:
```diff
@@ -742,6 +764,15 @@ static PyMethodDef methods[] = { // NOLINT
+    {"_get_multithreading_enabled",
+     get_multithreading_enabled,
+     METH_NOARGS,
+     nullptr},
```
Do this with a pybind binding on a guard class? It's a lot more compact to write.

ezyang commented on torch/csrc/autograd/init.cpp:
```diff
@@ -742,6 +764,15 @@ static PyMethodDef methods[] = { // NOLINT
+     get_multithreading_enabled,
+     METH_NOARGS,
+     nullptr},
+    {"_set_", set_grad_enabled, METH_O, nullptr},
```
hm?

albanD replied:
This needs to go I guess?

eellison replied:
oop

ezyang commented:
This looks basically fine but deferring to @albanD for final review.

albanD commented on torch/autograd/grad_mode.py:
```diff
@@ -298,3 +298,39 @@ def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+    .. note::
+        This API does not apply to :ref:`forward-mode AD <forward-mode-ad>`.
+
+    Example::
```
That can go yes.

albanD commented on c10/core/AutogradState.h:
```diff
@@ -41,10 +50,15 @@ struct C10_API AutogradState {
   bool grad_mode_ : 1;
   bool inference_mode_ : 1;
   bool fw_grad_mode_ : 1;
+  bool mulithreading_enabled_ : 1;
```
I'm curious if you really want this to be thread local. 
Aren't some of the things you want to catch is when you end up in another thread but the ThreadLocalState wasn't propagated?

eellison replied:
In the context of torch deploy, IMO it makes more sense for this to be thread local. And, if this is actually being respected in the backwards, we shouldn't have the problem of copying over TLS. But open to suggestions

albanD replied:
The TLS access shouldn't be too much of a perf regression here. So that's fine by me. Just wanted to double check.

albanD reviewed (approved):
Sounds good to me.

eellison committed:
Update on "Add Context Manager for Disabling Multithreading in Backwards, use in aot autograd"


We were running into a few issues with running multithreaded backwards in aot_autograd: such as https://github.com/pytorch/pytorch/issues/86136, and `FakeTensorMode` getting into a weird state as a result of not executing functions completely sequentially. The multithreaded backwards is lost in translation when we trace out the backwards anyway, and adds a lot of additional complexity. 


[ghstack-poisoned]

eellison committed:
Update on "Add Context Manager for Disabling Multithreading in Backwards, use in aot autograd"


We were running into a few issues with running multithreaded backwards in aot_autograd: such as https://github.com/pytorch/pytorch/issues/86136, and `FakeTensorMode` getting into a weird state as a result of not executing functions completely sequentially. The multithreaded backwards is lost in translation when we trace out the backwards anyway, and adds a lot of additional complexity. 


[ghstack-poisoned]

albanD commented on docs/source/torch.rst:
```diff
@@ -268,7 +268,7 @@ Examples::
     is_grad_enabled
     inference_mode
-    is_inference_mode_enabled
+    set_multithreading_enabled
```
This should be reverted.

eellison replied:
good catch, thanks

albanD reviewed (commented):
Small error in doc, good to go otherwise!

eellison committed:
Update on "Add Context Manager for Disabling Multithreading in Backwards, use in aot autograd"


We were running into a few issues with running multithreaded backwards in aot_autograd: such as https://github.com/pytorch/pytorch/issues/86136, and `FakeTensorMode` getting into a weird state as a result of not executing functions completely sequentially. The multithreaded backwards is lost in translation when we trace out the backwards anyway, and adds a lot of additional complexity. 


[ghstack-poisoned]

albanD reviewed (approved):
SGTM!

eellison committed:
Update on "Add Context Manager for Disabling Multithreading in Backwards, use in aot autograd"


We were running into a few issues with running multithreaded backwards in aot_autograd: such as https://github.com/pytorch/pytorch/issues/86136, and `FakeTensorMode` getting into a weird state as a result of not executing functions completely sequentially. The multithreaded backwards is lost in translation when we trace out the backwards anyway, and adds a lot of additional complexity. 


[ghstack-poisoned]

eellison commented:
@pytorchbot rebase

eellison committed:
Update on "Add Context Manager for Disabling Multithreading in Backwards, use in aot autograd"


We were running into a few issues with running multithreaded backwards in aot_autograd: such as https://github.com/pytorch/pytorch/issues/86136, and `FakeTensorMode` getting into a weird state as a result of not executing functions completely sequentially. The multithreaded backwards is lost in translation when we trace out the backwards anyway, and adds a lot of additional complexity. 


[ghstack-poisoned]

eellison commented:
@pytorchbot merge
